compiler/rustc_trait_selection/src/traits/query/normalize.rs

```diff
@@ -248,62 +257,61 @@ impl<'cx, 'tcx> FallibleTypeFolder<TyCtxt<'tcx>> for QueryNormalizer<'cx, 'tcx>
                 }
             }
 
-            ty::Alias(ty::Projection, data) if !data.has_escaping_bound_vars() => {
-                // This branch is just an optimization: when we don't have escaping bound vars,
-                // we don't need to replace them with placeholders (see branch below).
-
-                let tcx = self.infcx.tcx;
-                let data = data.try_fold_with(self)?;
+            ty::Opaque => ty.try_super_fold_with(self)?,
 
-                let mut orig_values = OriginalQueryValues::default();
-                // HACK(matthewjasper) `'static` is special-cased in selection,
-                // so we cannot canonicalize it.
-                let c_data = self
-                    .infcx
-                    .canonicalize_query_keep_static(self.param_env.and(data), &mut orig_values);
-                debug!("QueryNormalizer: c_data = {:#?}", c_data);
-                debug!("QueryNormalizer: orig_values = {:#?}", orig_values);
-                let result = tcx.normalize_projection_ty(c_data)?;
-                // We don't expect ambiguity.
-                if result.is_ambiguous() {
-                    // Rustdoc normalizes possibly not well-formed types, so only
-                    // treat this as a bug if we're not in rustdoc.
-                    if !tcx.sess.opts.actually_rustdoc {
-                        tcx.sess.delay_span_bug(
-                            DUMMY_SP,
-                            format!("unexpected ambiguity: {:?} {:?}", c_data, result),
-                        );
+            ty::Weak => {
+                let infcx = self.infcx;
+                let (data, maps) = if data.has_escaping_bound_vars() {
+                    let (data, mapped_regions, mapped_types, mapped_consts) =
+                        BoundVarReplacer::replace_bound_vars(infcx, &mut self.universes, data);
+                    (data, Some((mapped_regions, mapped_types, mapped_consts)))
+                } else {
+                    (data, None)
+                };
+                for (predicate, span) in
+                    infcx.tcx.predicates_of(data.def_id).instantiate_own(infcx.tcx, data.substs)
+                {
+                    let mut cause = self.cause.clone();
+                    cause.map_code(|code| ObligationCauseCode::TypeAlias(code, span, data.def_id));
+                    let obligation = Obligation::new(infcx.tcx, cause, self.param_env, predicate);
+                    // We don't bubble up obligations that are known to hold within the current param env.
+                    // `try_normalize_generic_arg_after_erasing_regions` expects only region obligations to be emitted.
+                    match infcx.evaluate_obligation(&obligation) {
```

Comment:
that should break if we have an obligation like `T: Trait<'a, 'b>` which requires `'b: 'a` becausewe return `EvaluatedToOkModuloRegions` in this case.

Probably easiest to change `normalize_projection_ty` to also be used here?

Reply:
same with `replace_bound_vars` here as for `project`

Reply:
> that should break if we have an obligation like `T: Trait<'a, 'b>` which requires `'b: 'a` becausewe return `EvaluatedToOkModuloRegions` in this case.
> 
> Probably easiest to change `normalize_projection_ty` to also be used here?

is this actually not an issue here or why did you resolve that comment?

Reply:
I forgot about that.

We don't even use `normalize_projection_ty` for opaque types yet. We should probably stop using `type_of(opaque)` and `type_of(weak)` to get the types behind the alias, and instead move all that logic into normalization.

That said... I am currently trying to build a repro first

Reply:
Yeah this doesn't seem right, hmmm.. Why was this marked as resolved?